### PR TITLE
Make date parsing more flexible for linedocsfile (europarl, enwiki)

### LINKED ISSUE #13073: java.lang.AssertionError in backward compat tests ("failed to parse ... as date")
### Description

These failures do reproduce but you need the linedocsfile. I'll take a look.

### Version and environment details

_No response_

### PR DESCRIPTION
**Pull request by dweiss**
Fixes #13073.

### CODE REVIEW

**Comment by dweiss** [comment]
This makes date parsing accept both europarl and enwiki. The tests still assume the docs come from europarl though and fail on the large enwiki.random.lines.txt:
```
gradlew -p lucene\backward-codecs test -Dtests.linedocsfile=/../enwiki.random.lines.txt
```

**Review by uschindler — APPROVED** [review]
Looks fine. Why did you create a `Function<String,LocalDateTime>` for parsing instead of a simple static method? I think you did this to hide the formatter instances?

File: lucene/backward-codecs/src/test/org/apache/lucene/backward_index/TestIndexSortBackwardsCompatibility.java [hunk 1]
```diff
@@ -147,6 +150,36 @@ public void testSortedIndex() throws Exception {
...
+            new DateTimeFormatterBuilder()
+                .parseStrict()
+                .parseCaseInsensitive()
+                .appendPattern("dd-MMM-uuuu HH:mm:ss['.'SSS]")
+                .toFormatter(Locale.ROOT);
+
+        @Override
+        public LocalDateTime apply(String s) {
+          if (s.matches("^[0-9]{4}-[0-9]{2}-[0-9]{2}$")) {
+            return euroParl.parse(s, LocalDate::from).atTime(LocalTime.MIDNIGHT);
```
Comment (uschindler) [inline lucene/backward-codecs/src/test/org/apache/lucene/backward_index/TestIndexSortBackwardsCompatibility.java:176]: `.atStartOfDay()` would be shorter.

**Comment by uschindler** [comment]
P.S.: Maybe add a quick test for the parsing logic to check that both formats are accepted.

**Comment by dweiss** [comment]
> Looks fine. Why did you create a `Function<String,LocalDateTime>` for parsing instead of a simple static method? I think you did this to hide the formatter instances?

Yes, correct.

**Comment by dweiss** [comment]
> P.S.: Maybe add a quick test for the parsing logic to check that both formats are accepted.

Maybe it'd be better to move this logic into LineFileDocs so that the date field's value is normalized somehow? Then we can add TestLineFileDocs and test it there. Adding a test method to TestIndexSortBackwardsCompatibility will multiply it over all the parameters, which doesn't make sense.

**Comment by dweiss** [comment]
I've moved that utility function to LineFileDocs and added a basic test case to TestLineFileDocs (good idea). I feel tempted to normalize the date field's value in LineFileDocs but this may break other stuff so better to leave it for a follow-up issue.

**Comment by dweiss** [comment]
I'll merge this in so that we can avoid jenkins failures. If there has to be a follow-up, I'll open another issue.

**Comment by s1monw** [comment]
LGTM thanks

**Pull request merged**